lib/segment/src/index/hnsw_index/graph_layers_builder.rs

```diff
@@ -234,23 +250,66 @@ impl GraphLayersBuilder {
         // ToDo: binary search here ? (most likely does not worth it)
         let new_to_target = score_internal(target_point_id, new_point_id);
 
-        let mut id_to_insert = links.len();
-        for (i, &item) in links.iter().enumerate() {
+        let mut id_to_insert = links.links.len();
+        for (i, &item) in links.links.iter().enumerate() {
             let target_to_link = score_internal(target_point_id, item);
             if target_to_link < new_to_target {
                 id_to_insert = i;
                 break;
             }
         }
 
-        if links.len() < level_m {
-            links.insert(id_to_insert, new_point_id);
-        } else if id_to_insert != links.len() {
-            links.pop();
-            links.insert(id_to_insert, new_point_id);
+        if links.links.len() < level_m {
+            links.links.insert(id_to_insert, new_point_id);
+        } else if id_to_insert != links.links.len() {
+            links.links.pop();
+            links.links.insert(id_to_insert, new_point_id);
         }
     }
 
+    fn select_one_candidate_with_heuristic_from_sorted<F>(
```

Comment:
What's the main difference with respect to `select_candidate_with_heuristic_from_sorted`? I may not be in the loop here but the naming stands out a little ambiguous.

Reply:
I tried to explain main idea of PR in notions
https://www.notion.so/qdrant/Heuristic-optimization-ee4ff1ee35cc4ce2828ef87d09dd300b